Piechart: Add visible keyboard focus indicator for data links

Abhijnya002 opened the pull request:
Fixes #114227

Adds visible focus indicator to data links in pie chart visualization to improve keyboard navigation accessibility.

- Added keyboard focus handlers for slices with data links
- Implemented highlight on focus using DataHoverEvent system
- Added Enter key support to activate data links
- Handled both single and multiple link cases
- Ensured proper tab order by sorting slices with data links first
- Added proper ARIA attributes (role, aria-label) for accessibility

**What is this feature?**

This PR adds keyboard accessibility support to pie chart data links. When users navigate using the Tab key, slices with data links now show a visible focus indicator (using the same highlight effect as mouse hover), making it clear which element is currently focused. Users can press Enter to activate data links via keyboard.

**Why do we need this feature?**

Currently, keyboard users cannot see which pie chart slice is focused when navigating with Tab, and cannot activate data links using the keyboard. This violates WCAG 2.4.7 (Focus Visible) accessibility guidelines and makes the pie chart inaccessible to keyboard-only users.

**Who is this feature for?**

- Keyboard-only users (accessibility requirement)
- Users who prefer keyboard navigation
- Screen reader users (with proper ARIA attributes)
- All users benefit from improved accessibility

**Which issue(s) does this PR fix?**:

Fixes #114227

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=115718) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=115718) before we can accept your contribution.<br/><hr/>**Abhijnya K G** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=115718) it.</sub>

Abhijnya002 committed:
fix(piechart): add visible keyboard focus indicator for data links

Fixes #114227

Adds visible focus indicator to data links in pie chart visualization
to improve keyboard navigation accessibility.

- Added keyboard focus handlers for slices with data links
- Implemented highlight on focus using DataHoverEvent system
- Added Enter key support to activate data links
- Handled both single and multiple link cases
- Ensured proper tab order by sorting slices with data links first
- Added proper ARIA attributes (role, aria-label) for accessibility

Abhijnya002 commented:
/add-changelog skip

gelicia commented on public/app/plugins/panel/piechart/PieChart.tsx:
```diff
@@ -115,40 +115,56 @@ export const PieChart = ({
-                  const color = arc.data.display.color ?? FALLBACK_COLOR;
-                  const highlightState = getHighlightState(highlightedTitle, arc);
+                {pie.arcs
+                  .sort((a, b) => {
```
I don't think we want to have the presence of data links affect pie slice order.

Abhijnya002 committed:
Remove data links sorting from pie chart slices

Remove the sort() call that was reordering pie slices based on data links presence. Pie slices should maintain their original order regardless of data links.

fastfrwrd commented on public/app/plugins/panel/piechart/PieChart.tsx:
```diff
@@ -243,15 +349,162 @@ function PieSlice({ arc, pie, highlightState, openMenu, fill, tooltip, tooltipOp
       onClick={openMenu}
+      tabIndex={shouldBeFocusable ? 0 : -1}
+      role={shouldBeFocusable ? 'link' : undefined}
+      aria-label={shouldBeFocusable ? `${arc.data.display.title} - Data link` : undefined}
```
two things:

1. this will need to be provided via `@grafana/i18n`, something like

```ts
t('piechart.data-link-label', '{{title}} - Data link', { title: arc.data.display.title })
```

2. no need to remove the label if `shouldBeFocusable` is false, we should just always add this label.

Abhijnya002 replied:
Fixed! The `aria-label` now uses `@grafana/i18n` and is always present.

fastfrwrd commented on public/app/plugins/panel/piechart/PieChart.tsx:
```diff
@@ -194,12 +198,114 @@ interface SliceProps {
+              },
+            });
+          }
+        };
```
```suggestion
        eventBus?.publish({
          type: DataHoverEvent.type,
          payload: {
            raw: e,
            x: 0,
            y: 0,
            dataId: arc.data.display.title,
          },
        });
```

fastfrwrd commented on public/app/plugins/panel/piechart/PieChart.tsx:
```diff
@@ -194,12 +198,114 @@ interface SliceProps {
+
+  useEffect(() => {
+    if (hasDataLinks && !openMenu && elementRef.current) {
+      const parentAnchor = elementRef.current.closest('a');
```
My main objection to this code is this big effect which attaches a several listeners to the `a` tag that may get created by `DataLinksContextMenu`. I think we should instead come up with a way to add these handlers directly to that component. There are situations which may be confusing and hard to test that could be introduced by us doing all of this handler management outside of the place where the component is created.

Abhijnya002 replied:
Fixed! Created a `PieChartDataLinksContextMenu` wrapper that handles the anchor event listeners where `DataLinksContextMenu` is used, removing the large useEffect from `PieSlice`. The handler logic is now co-located with component creation.

fastfrwrd commented on public/app/plugins/panel/piechart/PieChart.tsx:
```diff
@@ -243,15 +349,162 @@ function PieSlice({ arc, pie, highlightState, openMenu, fill, tooltip, tooltipOp
+            const actualX = svgRect.left + svgRect.width / 2 + centerX;
+            const actualY = svgRect.top + svgRect.height / 2 + centerY;
+
+            if (openMenu) {
```
while it's fundamentally a scope creep for this PR, I think the other biggest issue I have with this PR is right here. Since `openMenu` depends on the event being a mouse event right now, we can't trigger `openMenu` with a keypress, which is making it harder to add accessible handlers to components which use it, such as this one.

I think the right call is to do something like [this](https://github.com/grafana/grafana/pull/115903), where we update `DataLinksContextMenu` to allow either a keyboard or click event - in the case of a keyboard event, we'd need to get the x/y position using `getBoundingClientRect()`, as you have here, but I think we'd want to delegate that calculation down into the handler in the core component rather than doing it in pie chart. Please feel free to borrow that draft PR code into this one to improve it - I have not tested it with your changes yet because it requires a broader refactor of this code, but I did add tests which lead me to believe it would work.

Abhijnya002 replied:
Thanks for the feedback! I've extended `DataLinksContextMenu` to support keyboard events by:

1. Updated the API to accept position objects `{ x, y }` in addition to mouse events
2. Added `getMenuPosition` helper function that calculates position from an element using `getBoundingClientRect()`
3. Updated `WithContextMenu` to handle position objects and elements
4. Simplified the pie chart keyboard handler to use the new position object API

The position calculation logic is now in the core component (`DataLinksContextMenu`). The pie chart still has pie-specific logic for calculating from arc center, but it now uses the core component's position handling API.

This addresses the keyboard accessibility concern while keeping the position calculation in the core component as requested. Let me know if you'd like any adjustments!

fastfrwrd reviewed (changes requested):
Thanks for the PR! We do have a number of requests though, and many of them aren't the simplest to do unfortunately. Please let me know if you have any questions.

Abhijnya002 committed:
fix: use i18n for aria-label and always set it in pie chart

- Replace hardcoded aria-label string with @grafana/i18n translation
- Always set aria-label instead of conditionally setting to undefined
- Addresses review feedback on internationalization and accessibility

Abhijnya002 committed:
refactor: move outline style from inline to getSvgStyle CSS

- Move outline: 'none' from inline style prop to CSS styles
- Add outline: 'none' to all svgArg states (normal, highlighted, deemphasized)
- Addresses review feedback on code organization

Abhijnya002 commented on public/app/plugins/panel/piechart/PieChart.tsx:
```diff
@@ -243,15 +349,162 @@ function PieSlice({ arc, pie, highlightState, openMenu, fill, tooltip, tooltipOp
+      onKeyDown={shouldBeFocusable ? handleKeyDown : undefined}
+      onFocus={hasDataLinks ? handleFocus : undefined}
+      onBlur={hasDataLinks ? handleBlur : undefined}
+      style={{ outline: 'none' }}
```
Fixed! The outline style is now in `getSvgStyle` CSS instead of inline.

Abhijnya002 committed:
refactor: simplify ensureNotFocusable calls

- Remove unnecessary 100ms setTimeout call
- Keep immediate call and setTimeout(0) to handle React render timing
- Add comments explaining the rationale
- Addresses review feedback on code clarity

Abhijnya002 committed:
refactor: extract repeated event dispatch into useCallback

- Create publishDataHoverEvent and publishDataHoverClearEvent useCallback functions
- Replace all repeated eventBus.publish calls with the new callbacks
- Update dependency arrays to use the new callbacks
- Addresses review feedback on code duplication

Abhijnya002 commented on public/app/plugins/panel/piechart/PieChart.tsx:
```diff
@@ -194,12 +198,114 @@ interface SliceProps {
+              dataId: arc.data.display.title,
+            },
+          });
+        };
```
Done! I've extracted the repeated event dispatch code into two `useCallback` functions:
- `publishDataHoverEvent` for DataHoverEvent
- `publishDataHoverClearEvent` for DataHoverClearEvent

All handlers now reference these callbacks instead of duplicating the eventBus.publish code. This reduces code duplication and makes it easier to maintain.

Abhijnya002 committed:
refactor: move event handler logic into wrapper component

- Create PieChartDataLinksContextMenu wrapper that handles anchor event listeners
- Create PieSliceWithDataLinks component that uses the wrapper
- Remove large useEffect from PieSlice that was finding anchor after creation
- Event handler logic is now co-located with DataLinksContextMenu usage
- Addresses review feedback about handler management being outside component creation

Abhijnya002 committed:
feat: extend DataLinksContextMenu to support keyboard events

- Update DataLinksContextMenu API to accept position objects for keyboard events
- Add getMenuPosition helper function to calculate position from element
- Update WithContextMenu to handle position objects and elements
- Simplify pie chart keyboard handler to use position object API
- Position calculation logic now in core component (DataLinksContextMenu)
- Addresses review feedback about keyboard accessibility and position calculation

Abhijnya002 committed:
docs: add comments explaining blur timeout pattern

- Explain why blur timeout is needed for focus transitions
- Clarify that it handles focus moving between related UI elements
- Addresses review feedback requesting explanation of blur timeout

Abhijnya002 commented on public/app/plugins/panel/piechart/PieChart.tsx:
```diff
@@ -243,15 +349,162 @@ function PieSlice({ arc, pie, highlightState, openMenu, fill, tooltip, tooltipOp
+    (event: React.FocusEvent<SVGGElement>) => {
+      if (blurTimeoutRef.current) {
+        clearTimeout(blurTimeoutRef.current);
+        blurTimeoutRef.current = null;
```
The blur timeout is needed to handle focus transitions between related UI elements. When a context menu opens (via keyboard Enter or click), focus may move from the pie slice to the menu, triggering a blur event. Without the timeout, we'd immediately clear the hover state even though focus is just moving to a related element.

The 100ms delay allows us to check if focus has actually left the component or just moved to a related element. If focus returns to the element (or a related element) within 100ms, `handleFocus` clears the timeout, preventing the hover state from being cleared prematurely.

I've added comments to the code explaining this pattern. This is a common approach for handling focus transitions in complex UI components where focus might move between related elements.

fastfrwrd replied:
I buy this, thanks for putting it in a comment.

fastfrwrd commented on packages/grafana-ui/src/components/ContextMenu/WithContextMenu.tsx:
```diff
@@ -15,16 +15,38 @@ export interface WithContextMenuProps {
-            y: e.pageY - window.scrollY,
-          });
-        },
+        openMenu: handleOpenMenu as React.MouseEventHandler<HTMLElement>,
```
can we figure out any way to get rid of the cast? we should centralize and export the definition of this callback type, and then use that type instead of the constituent types.

```ts
export type WithContextMenuOpenMenu = 
  (e: React.MouseEvent<HTMLElement | SVGElement> | React.SyntheticEvent<HTMLElement | SVGElement> | { x: number; y: number }) =>  void;
```

fastfrwrd reviewed (commented):
I spent about an hour today working with this branch, and I still have some issues with how this works that I want to address before we move forward with this.

The highlights, to me:

- I think we can cut down on the anti-patterns we need to get this done still. there's a lot of cruft with the TypeScript types for example, which we can overcome by defining a core type for the `openMenu` callback method. I also want to minimize the number of signatures a little bit, so that's why I went with targeting `e.currentTarget` instead of just passing in an element.
- I still don't think we should do the parentAnchor stuff this way, particularly the event binding. I want to explore other ways to set that up that don't feel so brittle.
- we should add e2e tests for all of these scenarios to make sure they continue to work (there's not much in our pie-chart e2es yet), and I want to add unit tests for the WithContextMenu component for sure.

I created a [branch off of yours](https://github.com/grafana/grafana/tree/fastfrwrd/fix/piechart-keyboard-focus-indicator) where I made some progress on these items, feel free to use it and keep working. I think the biggest challenge remaining in my branch is the `parentAnchor`/SVG event stuff - when I tested it today, pie slices weren't even rendering when I set one or more datalinks.

Abhijnya002 committed:
refactor: simplify context menu types and improve pie chart data links

- Define WithContextMenuOpenMenuCallback type to reduce TypeScript complexity
- Use e.currentTarget pattern instead of passing elements directly
- Replace brittle querySelector approach with ref callback for anchor events
- Simplify DataLinksContextMenu to use the new callback type
- Clean up pie chart implementation to use improved patterns

Abhijnya002 committed:
test: add unit tests for WithContextMenu and e2e tests for pie chart keyboard navigation

- Add unit tests for WithContextMenu component covering mouse, keyboard, and position object scenarios
- Add e2e tests for pie chart keyboard navigation with data links
- Test both single-link and multiple-links scenarios
- Verify focus behavior and context menu interaction

Abhijnya002 commented:
Thanks for the feedback. Addressed all the points:
1. Simplified TypeScript types:
- Added WithContextMenuOpenMenuCallback as the core type for openMenu
- Updated DataLinksContextMenu to use this type, removing complex union types
- Reduced signature complexity
2. Used e.currentTarget pattern:
- Updated WithContextMenu to use e.currentTarget.getBoundingClientRect() for keyboard events
- Removed element parameter passing
- Simplified the API
3. Replaced brittle parentAnchor approach:
- Replaced querySelector with a React ref callback for anchor events
- Extended DataLinksContextMenu to accept anchor event handler props (anchorRef, onAnchorFocus, etc.)
- Moved event handler logic into PieSliceWithDataLinks using ref callbacks
- More React-friendly and less fragile
4. Fixed rendering issue:
- Tested with single and multiple data links; pie slices render correctly
- Verified keyboard navigation (Tab, Enter, Space) works for both scenarios
- Confirmed structure: single link uses anchor wrapper, multiple links use SVG directly
5. Added tests:
- Unit tests for WithContextMenu component covering mouse events, keyboard events (SyntheticEvent), position objects, and edge cases
- E2e tests for pie chart keyboard navigation including single-link anchor focus, multiple-links context menu interaction, and keyboard navigation scenarios

All requested items are addressed. Ready for review.

Abhijnya002 commented:
Thanks again - all requested updates are in place.
